### DISCUSSION TITLE
install manually

### DISCUSSION THREAD

**Discussion by mastermmz** [serverpod/serverpod#2067] [Q&A] [answered]
The following command takes a long time for me to execute:
$ dart pub global activate serverpod_cli

Is there a way to download the files from GitHub and install serverpod manually?

**Comment by SandPod** [comment, marked as answer]
Hey @mastermmz 👋 

We don't store a binary with the serverpod cli but there are two ways to run the cli straight from GitHub instead of going through pub dev.

The following steps only work if you are on a Unix system (MacOS or Linux).

Either you can run it in developer mode where all packages will be referenced locally through the following steps:
1. Clone our repo from a stable release tag (if you want to experiment with the latest you can pick `main` or a stable branch).
2. Navigate to the cloned directory and run `./util/pub_get_all`.
3. Run the command: `SERVERPOD_HOME=$(pwd) && dart pub global activate --source path tools/serverpod_cli`
4. Run the `serverpod version` to validate that the installation is successful.

The `SERVERPOD_HOME` environment variable is used to resolve the path to our templates.

Or you can compile the production templates and have the project look like it was downloaded from pub dev:
1. Clone our repo from a stable release tag
2. Navigate to the cloned directory and run `./util/pub_get_all`
3. Run `dart tools/serverpod_cli/bin/serverpod_cli.dart generate-pubspecs --version <TAG VERSION> --mode production` to compile the production templates
5. Run `SERVERPOD_HOME=$(pwd) && dart pub global activate --source path tools/serverpod_cli` to activate the CLI.
6. Then run the command `serverpod version` to validate that the installation was successful.

**Reply by AndryHTC** [reply]
Any suggestions on how to make a custom Serverpod work on Windows 11?

I've been able to convert the commands into PowerShell, and it seemed to be on the right track when I used `serverpod generate`. I’m trying to open a pull request to Serverpod, so I need the local version working. I’ve changed some code in the repo, successfully converted the commands, and was able to run `serverpod generate` once.

However, after rolling back the changes, the generation still gives me the same result as when the changes were applied. What am I doing wrong? Do I need to run something specific to ensure Serverpod works with the updated files? I’ve tried re-running all the commands, but nothing seems to fix it.

**Reply by SandPod** [reply]
@AndryHTC - Everytime you make changes to the Serverpod CLI you need to reactivate the cli.

This triggers a rebuild of the cli binary with your new changes.
So running `dart pub global activate --source path tools/serverpod_cli` should solve your problem.

If you still get the same result, we have noticed that dart sometimes does not correctly remove the binary that was built.
Then, you manually need to remove it before activating the CLI.

On Mac, I can find out where the cached binary is stored by running:
```
$ which serverpod
-> /Users/alex/.pub-cache/bin/serverpod
$ cat /Users/alex/.pub-cache/bin/serverpod
-> 
#!/usr/bin/env sh
# This file was created by pub v3.5.1.
# Package: serverpod_cli
# Version: 2.1.0-beta.3
# Executable: serverpod
# Script: serverpod_cli
if [ -f /Users/alex/.pub-cache/global_packages/serverpod_cli/bin/serverpod_cli.dart-3.5.1.snapshot ]; then
  dart "/Users/alex/.pub-cache/global_packages/serverpod_cli/bin/serverpod_cli.dart-3.5.1.snapshot" "$@"
  # The VM exits with code 253 if the snapshot version is out-of-date.
  # If it is, we need to delete it and run "pub global" manually.
  exit_code=$?
  if [ $exit_code != 253 ]; then
    exit $exit_code
  fi
  dart pub -v global run serverpod_cli:serverpod_cli "$@"
else
  dart pub global run serverpod_cli:serverpod_cli "$@"
fi
```

And then manually remove the file `/Users/alex/.pub-cache/global_packages/serverpod_cli/bin/serverpod_cli.dart-3.5.1.snapshot`

**Reply by AndryHTC** [reply]
Great. My intuition was right so. But really... this DX got me a couple of monitors and keyboards shattered

**Reply by AndryHTC** [reply]
Done a `dart pub cache clean` and re-runned everything. I still do not get the fresh binary.

It doesn't even seem to be generated in the `pub/cache/global_packages/serverpod_cli` folder.

Trying with my Macbook now, but a 2h job became a 3 days one

**Reply by AndryHTC** [reply]
```
andry@MBP-di-Andrea serverpod % SERVERPOD_HOME=$(pwd) && dart pub global activate --source path tools/serverpod_cli     
Resolving dependencies in `tools/serverpod_cli`... 
Downloading packages... 
  _fe_analyzer_shared 73.0.0 (74.0.0 available)
  analyzer 6.8.0 (6.9.0 available)
  macros 0.1.2-main.4 (0.1.3-main.0 available)
! serverpod_client 2.1.0-beta.3 from path packages/serverpod_client (overridden)
! serverpod_lints 2.1.0-beta.3 from path packages/serverpod_lints (overridden)
! serverpod_serialization 2.1.0-beta.3 from path packages/serverpod_serialization (overridden)
! serverpod_service_client 2.1.0-beta.3 from path packages/serverpod_service_client (overridden)
! serverpod_shared 2.1.0-beta.3 from path packages/serverpod_shared (overridden)
Got dependencies in `tools/serverpod_cli`!
3 packages have newer versions incompatible with dependency constraints.
Try `dart pub outdated` for more information.
Package serverpod_cli is currently active at path "/Users/andry/Development/serverpod/tools/serverpod_cli".
Installed executable serverpod.
Activated serverpod_cli 2.1.0-beta.3 at path "/Users/andry/Development/serverpod/tools/serverpod_cli".
```

```
andry@MBP-di-Andrea serverpod % serverpod version
Building package executable... (2.9s)
Built serverpod_cli:serverpod_cli.
The SERVERPOD_HOME environmental variable is required in development mode,
```

This is a fresh install in a MBP. I'm done

**Reply by SandPod** [reply]
Given the last print you are almost there, try: `SERVERPOD_HOME=<path-to-serverpod-repo-root> serverpod version`.

The `SERVERPOD_HOME` env needs to be present when running commands with your local Serverpod CLI. You can export it to save it between runs.

`export SERVERPOD_HOME=<path-to-serverpod-repo-root>`

**Reply by SandPod** [reply]
This is documented in our contribution guidelines here: https://docs.serverpod.dev/next/contribute#command-line-tools

**Reply by AndryHTC** [reply]
@SandPod I wanted to try with a Mac to be sure wasn't a Windows specific issue.

> Done a `dart pub cache clean` and re-runned everything. I still do not get the fresh binary.
> 
> It doesn't even seem to be generated in the `pub/cache/global_packages/serverpod_cli` folder.
> 
> Trying with my Macbook now, but a 2h job became a 3 days one
